### ISSUE TITLE
Rule Proposal: stateless-function-component or no-stateless-component-class

### ISSUE THREAD

**Issue by oguzbilgic** [yannickcr/eslint-plugin-react#214]
Rule would warn if a stateless component isn't written as a pure function

**Comment by lo1tuma** [comment]
Maybe `prefer-stateless-function` would be a better name.

**Comment by yannickcr** [comment]
Can you give me some examples of what you would like to forbid/validate with this rule ?

**Comment by lo1tuma** [comment]
The rule should warn for components that are defined as a class or with `React.createClass` but could be safely written as pure stateless functions.

This should be considered as a problem:

```jsx
class Foo extends React.Component {
    render() {
        return <div>{this.props.foo}</div>;
    }
}
```

This should not be considered as a problem:

```jsx
const Foo = function (props) {
     return <div>{props.foo}</div>;
};
```

```jsx
const Foo = ({foo}) => <div>{foo}</div>;
```

```jsx
// using other lifecycle methods than render
class Foo extends React.Component {
    shouldComponentMount() {
        return fasle;
    }

    render() {
        return <div>{this.props.foo}</div>;
    }
}
```

```jsx
// using state
class Foo extends React.Component {
    changeState() {
        this.setState({foo: 'clicked'});
    }

    render() {
        return <div onClick={this.changeState.bind(this)}>{this.state.foo || 'bar'}</div>;
    }
}
```

**Comment by oguzbilgic** [comment]
@gaearon You might wanna chime in to this, based on your tweet.

https://twitter.com/dan_abramov/status/656585973552324608

**Comment by gaearon** [comment]
I agree that's a cool thing to have.
I don't have time or expertise to help with implementation at the moment.

**Comment by yannickcr** [comment]
Working on it.

**Comment by gabrielenosso** [comment]
How should a component with own methods (which are not lifecycle methods) be?

Example:
`class Foo extends React.Component {
    logout() {
        this.props.actions.logout();
    }

    render() {
        return <div onClick={this.logout}>Yeah!</div>;
    }
}`

**Comment by ljharb** [comment]
@yannickcr it's possible the rule has a bug here ^ if a component has instance methods, or instance properties attached in the constructor, it shouldn't be an SFC, because it has state - `this` is state.

**Comment by yannickcr** [comment]
@ljharb Yes, I responded in https://github.com/yannickcr/eslint-plugin-react/issues/491#issuecomment-193954254. I think a check is missing for `this`.

**Comment by gabrielenosso** [comment]
Any news about this? Will it be fixed in next release? :)

Anyway, thanks for your great work!

**Comment by yannickcr** [comment]
@gabrielenosso I should work on it this week-end, you can subscribe to #491 to track issue progress.
